File: LeetCode/Medium/1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int n = nums.size();
        int ret = 0;
        int l = 0;
        int r = 0;
        
        while (r < n){
            while (r < n && k > 0) {
                if (nums[r] == 0) k--;
                ++r;
            }
            
            while (r < n && nums[r] == 1) ++r;
            
            ret = max(ret, r - l);
            
            while (l < n && nums[l] == 1) ++l;
            
            ++l;
            ++k;
        }
        
        return ret;
    }
};

File: LeetCode/Medium/1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        ret, l, r = 0, 0, 0

        while r < len(nums):
            while r < len(nums) and k:
                if nums[r] == 0:
                    k -= 1
                r += 1

            while r < len(nums) and nums[r] == 1:
                r += 1

            ret = max(ret, r - l)

            while l < len(nums) and nums[l] != 0:
                l += 1
                
            l += 1
            k += 1

        return ret